### PR TITLE
[Core] Improve coupling geometry

### PR DESCRIPTION
**Pull request by loumalouomega**
This is a transition PR. I am planning to use coupled geometries for contact instead of the current approach where I store the indexes of the coupled conditions

This adds RemoveGeometryPart and makes virtual the Set and Add methods

### CODE REVIEW

**Comment by loumalouomega** [comment]
@roigcarlo do you know why it fails in Windows?

**Comment by roigcarlo** [comment]
Nop, to be honesest the exit code looks odd. Maybe was a problem in the CI. try to recompile and let's see whats happens if it keeps failing with the same code I will try to compile it locally.

**Comment by loumalouomega** [comment]
I already recompiled and does not work. Unfortunately I don't use Windows,
can anyone check it?

**Comment by tteschemacher** [comment]
Hey @loumalouomega thanks a lot for the changes. On Tuesday I can have a look at the windows error massage. Besides, I'm totally fine with the changes. But usually @pooyan-dadvand reviewed the according changes in the geometry.h.

Just for my information, does the const IndexType change anything?

**Comment by loumalouomega** [comment]
In theory the compiler optimises more (and least in some types)

File: kratos/geometries/coupling_geometry.h [hunk 1]
```diff
@@ -268,13 +270,55 @@ template<class TPointType> class CouplingGeometry
...
+    }
+
+    /**
+     * @brief Use to check if certain Indexed object is within the geometry parts of this geometry.
+     * @param Index of the geometry part. This index can be used differently within the derived classes.
+     * @return true if has geometry part
+     */
+    bool HasGeometryPart(const IndexType Index) const override
+    {
+        if (Index <= NumberOfGeometryParts() - 1) {
```
Comment (tteschemacher) [inline kratos/geometries/coupling_geometry.h]:
```suggestion
        if (Index < NumberOfGeometryParts()) {
```

**Comment by pooyan-dadvand** [comment]
@loumalouomega seems that the structural mechanic is not finishing. Would you please increase the verbosity of the test to see which of them is failing. And would you please check if Valgrind is clean?

IMO, the problem is concurrency in openMP. So you may try to put that test in 1 thread to see if it passes

**Comment by loumalouomega** [comment]
@tteschemacher looks like I fix the error in windows

**Comment by tteschemacher** [comment]
nice. was the error related to those code changes or was is coming from the master?

**Review by tteschemacher — APPROVED** [review]
I am fine with the changes. Please @pooyan-dadvand give your feedback.

**Comment by loumalouomega** [comment]
The error was because the implementation of the erase in the std::vector in
Windows compiler. I did what the erase does in the std by hand and solved
the problem

**Comment by tteschemacher** [comment]
thanks, because I have a similar problem in #7047 and I don't know how to solve it.

**Comment by loumalouomega** [comment]
Do you agree @pooyan-dadvand ?

**Comment by loumalouomega** [comment]
> Do you agree @pooyan-dadvand ?

Do you agree @pooyan-dadvand ?

**Comment by tteschemacher** [comment]
It looks nice. Thanks @loumalouomega . @pooyan-dadvand this is what we discussed some days, ago.

Btw. I made a bit of research. Making the index const doesn't change anything in the performance. I'm fine with having it but it shouldn't affect anything as it's a copy.

**Pull request merged**